### PR TITLE
Fixes #28564 - improve performance of orphaned facts cleanup

### PR DESCRIPTION
**Pull request by tbrisker**
This should greatly improve the performance of cleaning up orphaned
facts. First, we stop trying to delete fact_values for orphaned facts
since they don't exist. Then, instead of checking each compose fact's
descendants to see if they were all deleted, we get a set of all
remaining leaf's ancestors and delete any not on that list.

### CODE REVIEW

**Comment by tbrisker** [comment]
cleaning up 64K composed facts with no children took around 6 minutes with this patch, where previously the task was running for hours in the delete_orphaned_facts step, taking around 4 minutes for each batch of 500 to be processed.

File: app/services/fact_cleaner.rb [hunk 1]
```diff
@@ -32,17 +32,19 @@ def delete_facts_names_values(fact_name_ids)
...
-      leaf_ids += fact_name_ids
+      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
     end
 
     logger.debug "Cleaning compose orphaned facts"
-    FactName.composes.reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |compose| compose.descendants.all? { |fact| leaf_ids.include?(fact.id) } }.map(&:id)
-      @deleted_count += delete_facts_names_values(fact_name_ids)
+    live_ancestors = Set.new
+    FactName.leaves.reorder(nil).select(:ancestry, :id).find_each(:batch_size => @batch_size) do |leaf|
```
Comment (ShimShtein) [inline app/services/fact_cleaner.rb]: Instead of `:ancestry` you can use `FactName.ancestry_base_class.ancestry_column`.

**Review by ShimShtein — COMMENTED** [review]
LGTM. Small change to be more resilient to DB changes in the future, but it's not necessary.

File: app/services/fact_cleaner.rb [hunk 2]
```diff
@@ -32,17 +32,19 @@ def delete_facts_names_values(fact_name_ids)
...
+      live_ancestors.merge(leaf.ancestor_ids)
+    end
+    FactName.composes.reorder(nil).where.not(id: live_ancestors).select(:id).find_in_batches(:batch_size => @batch_size) do |batch|
```
Comment (lzap) [inline app/services/fact_cleaner.rb:45]: I like `select(:id)` can you also add it to the other queries?

**Review by lzap — CHANGES_REQUESTED** [review]
Yeah nice one.

File: app/services/fact_cleaner.rb (hunk 1 above)
Comment (tbrisker) [reply app/services/fact_cleaner.rb]: updated

File: app/services/fact_cleaner.rb (hunk 2 above)
Comment (tbrisker) [reply app/services/fact_cleaner.rb:45]: added to the pattern one, not sure if it will work well with the preload for the leafs one

**Comment by lzap** [comment]
What's up with "SmartProxiesControllerTest.test_0009_smart proxy version with previous y warns" error? Is this broken develop?

**Comment by tbrisker** [comment]
> What's up with "SmartProxiesControllerTest.test_0009_smart proxy version with previous y warns" error? Is this broken develop?

yes, looks like i broke develop with the 2.0 version bump

File: app/services/fact_cleaner.rb [hunk 3]
```diff
@@ -32,25 +32,27 @@ def delete_facts_names_values(fact_name_ids)
...
 
     logger.debug "Cleaning compose orphaned facts"
-    FactName.composes.reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |compose| compose.descendants.all? { |fact| leaf_ids.include?(fact.id) } }.map(&:id)
-      @deleted_count += delete_facts_names_values(fact_name_ids)
+    live_ancestors = Set.new
+    FactName.leaves.reorder(nil).select(FactName.ancestry_base_class.ancestry_column, :id).find_each(:batch_size => @batch_size) do |leaf|
+      live_ancestors.merge(leaf.ancestor_ids)
+    end
+    FactName.composes.reorder(nil).where.not(id: live_ancestors).select(:id).find_in_batches(:batch_size => @batch_size) do |batch|
```
Comment (ShimShtein) [inline app/services/fact_cleaner.rb:45]: We can improve it even further by using `.in_batches.pluck` here. We don't need the model object at all

Comment (lzap) [reply app/services/fact_cleaner.rb:45]: I had no idea this exists.

https://github.com/rails/rails/blob/v5.0.0/activerecord/lib/active_record/relation/batches.rb#L184

This is a very good memory optimization which is worth looking into - array of integers will be much more effective than AR instances. Could you incorporate this change? Can you also optimize all the other statements? Should be pretty straightforward change.

However if you don't want to do this it's fine.

Comment (tbrisker) [reply app/services/fact_cleaner.rb:45]: i tired that now and looks like it actually doubles the queries- it does one query for the ids to get the batch, then another to pluck the id from all the records in the batch

Comment (tbrisker) [reply app/services/fact_cleaner.rb:45]: specifically this is the code: https://github.com/rails/rails/blob/v5.0.0/activerecord/lib/active_record/relation/batches.rb#L205-L206 
doesn't look like there is an easy way of getting just the ids without doing our own batch handling.

**Review by lzap — APPROVED** [review]
Thanks.

**Comment by tbrisker** [comment]
1.24 - aae7748e19231a983e52e29b7a56254655f5cf60

**Pull request merged**